src/core_plugins/input_control_vis/public/vis_controller.js

```diff
@@ -0,0 +1,113 @@
+import _ from 'lodash';
+import React from 'react';
+import { render, unmountComponentAtNode } from 'react-dom';
+import { InputControlVis } from './components/vis/vis';
+import { controlFactory } from './control/control_factory';
+
+class VisController {
+  constructor(el, vis) {
+    this.el = el;
+    this.vis = vis;
+    this.controls = [];
+
+    this.queryBarUpdateHandler = this.updateControlsFromKbn.bind(this);
+    this.vis.API.queryFilter.on('update', this.queryBarUpdateHandler);
+  }
+
+  render(visData, status) {
+    if (status.params) {
+      this.initControls();
+    }
+    return new Promise(() => {});
+  }
+
+  resize() {}
+
+  destroy() {
+    this.vis.API.queryFilter.off('update', this.queryBarUpdateHandler);
+    unmountComponentAtNode(this.el);
+  }
+
+  drawVis() {
+    render(
+      <InputControlVis
+        updateFiltersOnChange={this.vis.params.updateFiltersOnChange}
+        controls={this.controls}
+        stageFilter={this.stageFilter.bind(this)}
+        submitFilters={this.submitFilters.bind(this)}
+        resetControls={this.updateControlsFromKbn.bind(this)}
+        clearControls={this.clearControls.bind(this)}
+        getStagedControls={this.getStagedControls.bind(this)}
+      />,
+      this.el);
+  }
+
+  initControls() {
+    this.controls = [];
+
+    controlFactory(
```

Comment:
I'm thinking that if we change the `controlFactory` to deal with a single `control`, then we could switch from using a `callback` below for each individual control and use a `Promise` which seems more natural.

This would also allow us to use a `Promise.all` for each of the controlFactory to use with the `render` method.

Reply:
I did it this way so that a single `_msearch` would be used to init all controls in a panel. The weirdness comes from the fact that `fetch.these` uses the defer promise pattern